### PR TITLE
FermionOperatorJax

### PR DESCRIPTION
**Pull request by PhilipVinc**

### CODE REVIEW

File: netket/experimental/operator/_fermion_operator_2nd_base.py [hunk 1]
```diff
@@ -47,9 +45,18 @@
 )
 
 
-class FermionOperator2nd(DiscreteOperator):
+# TODO is this a reasonable cutoff?
+@numba.jit(nopython=True)
+def _isclose(a, b, cutoff=1e-6):  # pragma: no cover
+    return np.abs(a - b) < cutoff
```
Comment (inailuig) [inline netket/experimental/operator/_fermion_operator_2nd_base.py:51]:
```suggestion
```

File: netket/experimental/operator/_fermion_operator_2nd_base.py [hunk 2]
```diff
@@ -539,7 +368,7 @@ def _op__imatmul__(self, other):
         return self
 
     def _op__matmul__(self, other):
-        if not isinstance(other, FermionOperator2nd):
+        if not isinstance(other, FermionOperator2ndBase):
```
Comment (inailuig) [inline netket/experimental/operator/_fermion_operator_2nd_base.py:371]:
```suggestion
        if not isinstance(other, type(self)):
```
Do we really want to allow adding numba + jax ?

(same for for add below)

File: netket/experimental/operator/_fermion_operator_2nd_jax.py [hunk 3]
```diff
@@ -0,0 +1,686 @@
...
+                weight_dtype=self._dtype,
+            )
+
+            # TODO the following could be reduced further
+            self._max_conn_size = int(len(self._terms_list_diag) > 0) + len(
+                offdiag_operators
+            )
+            self._initialized = True
+
+            self._kwargs = {"apply_terms_fun": apply_terms_scan_bits}
```
Comment (inailuig) [inline netket/experimental/operator/_fermion_operator_2nd_jax.py:609]: I had introduced this field is as a hacky way to set the implementation. 
Maybe we could add the constructor to this class with a private parameter for it, similar to `_mode` in paulistringsjax.

IIrc apply_terms_scan_bits was the fastest implementation, which is why it is set as default here.
(this is memory bound, so doing the computation with bitmasks helps)
One day we should do some proper benchmarking...

Also I think the one with masks might be useful for people which want to change their parameters without recompiling.

File: netket/experimental/operator/_fermion_operator_2nd_jax.py [hunk 4]
```diff
@@ -0,0 +1,686 @@
...
+@partial(jax.vmap, in_axes=(None, 0, 0, 0, None), out_axes=(-2, -1, -1))
+def _apply_terms_scan_math(x, w, sites, daggers, unroll):
+    return _apply_term_scan_math(x, w, sites, daggers, unroll=unroll)
+
+
+@partial(jax.jit, static_argnums=(4, 5))
+def apply_terms_scan_math(x, w, sites, daggers, unroll=1, dtype=jnp.float32):
+    # TODO avoid casting here, return orig type then
+    x = x.astype(dtype)
+    return _apply_terms_scan_math(x, w, sites, daggers, unroll)
```
Comment (inailuig) [inline netket/experimental/operator/_fermion_operator_2nd_jax.py:202]:
```suggestion
```
this was an experiment I did for gpu (I had in mind of using half or even lower precision),
iirc it didn't perform much better, and we can probably remove this one, in favour of the other implementations.

File: netket/experimental/operator/_fermion_operator_2nd_base.py (hunk 1 above)
Comment (inailuig) [reply netket/experimental/operator/_fermion_operator_2nd_base.py:51]: Seems like an artifact of the rebase. I would remove numba from the base class altogether, and use `np.isclose` instead of this.

File: netket/experimental/operator/_fermion_operator_2nd_jax.py (hunk 3 above)
Comment (PhilipVinc) [reply netket/experimental/operator/_fermion_operator_2nd_jax.py:540]: the implementation is `apply_terms_masks`or `apply_terms_only_masks`?

File: netket/experimental/operator/_fermion_operator_2nd_base.py (hunk 2 above)
Comment (PhilipVinc) [reply netket/experimental/operator/_fermion_operator_2nd_base.py:364]: I think it's ok for now. At least until we allow for manipulating them inside jit expressions (which we don't yet) this should not lead to any problems...

File: netket/experimental/operator/_fermion_operator_2nd_jax.py [hunk 5]
```diff
@@ -0,0 +1,663 @@
...
+
+@register_pytree_node_class
+class FermionOperator2ndJax(FermionOperator2ndBase, DiscreteJaxOperator):
+    r"""
+    A fermionic operator in :math:`2^{nd}` quantization using pure jax dark
+    magic for indexing.
+
+    .. warning::
+
+        This class is not a Pytree, so it **can** be used inside of jax-transformed
```
Comment (gcarleo) [inline netket/experimental/operator/_fermion_operator_2nd_jax.py:498]: is not a Pytree?

Comment (PhilipVinc) [reply netket/experimental/operator/_fermion_operator_2nd_jax.py:498]: typo

File: netket/experimental/operator/_fermion_operator_2nd_jax.py [hunk 6]
```diff
@@ -0,0 +1,663 @@
...
+
+
+# mostly masks, some indexing
+def _apply_term_masks(x, w, sites, daggers):
+    # sites can be an unsigned int
+    # daggers and x need to be signed, preferably of the same type
+
+    # assert x.dtype == np.int8
+    # assert daggers.dtype == np.int8
+    # assert sites.dtype == np.uint8
```
Comment (gcarleo) [inline netket/experimental/operator/_fermion_operator_2nd_jax.py:253]: so you are saying we can do maximum 256 sites? :)

Comment (PhilipVinc) [reply netket/experimental/operator/_fermion_operator_2nd_jax.py:253]: it's commented..

Comment (gcarleo) [reply netket/experimental/operator/_fermion_operator_2nd_jax.py:253]: but there are other places where you actually use np.int8 ...

Comment (PhilipVinc) [reply netket/experimental/operator/_fermion_operator_2nd_jax.py:253]: For `x` and `dagger` which are both boolean variables (the first is the bit string of occupation numbers (0,1) and the latter is a boolean saying if it's creation or destruction.

It's not used for sites

File: test/operator/test_operator.py [hunk 7]
```diff
@@ -110,6 +110,22 @@ def _loc(*args):
...
+    weights=(0.5 + 0.3j, 0.5 - 0.3j),  # must add h.c.
+)
+
+# TODO: @inailuig Fix this broken test
+# operators["FermionOperator2ndJax(_mode=mask)"] = nkx.operator.FermionOperator2ndJax(
+#    hi,
+#    terms=(((0, 1), (3, 0)), ((3, 1), (0, 0))),
+#    weights=(0.5 + 0.3j, 0.5 - 0.3j),  # must add h.c.
+#    _mode="mask"
+# )
```
Comment (PhilipVinc) [inline test/operator/test_operator.py:127]: @inailuig the mode mask is broken. I commented the test, but if at some point you have some time it would be lovely if you could fix this.

Regardless I'll merge this as is.

**Comment by jwnys** [comment]
What is still needed to move fermions out of experimental?

**Comment by PhilipVinc** [comment]
I'm in favour of moving fermions out of experimental.

Would love the new sampler to be in experimental or in an unreleased version for at least 2 months to gather feedback.

The only thing that I would like discuss is how to move them out of experimental. 
We can simply move them out to the main namespace, but maybe we should have a fermion namespace? If we start to have fermion-specific samplers and ansatze...

Or maybe this is not needed, and we can just put them in the main namespace and take care to separate them in the documentation.

**Comment by jwnys** [comment]
Ok, let me remove this annoying constant everywhere and make the cutoff explicit everywhere (the latter might give inconsistencies at the moment).
I would not move it to a specific namespace personally..

**Comment by jwnys** [comment]
See #1686
